src/main/java/com/google/sps/data/GivrUser.java

```diff
@@ -51,53 +55,119 @@ public boolean isMaintainer() {
+      throw new IllegalArgumentException("User with " + identifyingProperty + " of value: " + identifyingValue + " was not found in the Datastore.");
+    }
 
-    if (userResult.size() == 1) {
-      for (Entity entity: preparedQuery.asIterable(fetchOptions)) {
-        isMaintainer = (boolean) entity.getProperty("isMaintainer");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery preparedQuery = getPreparedQueryResultOfUserWithProperty(identifyingProperty, identifyingValue);
+    Entity entity = preparedQuery.asSingleEntity();
+
+    if (entity != null) {
+      Transaction txn = datastore.beginTransaction();
+      try {
+        for (Map.Entry<String, String> entry: updatePropertyNamesAndValues.entrySet()) {
+          entity.setProperty(entry.getKey(), entry.getValue());
+        }
+
+        datastore.put(txn, entity);
+        txn.commit();
+      } finally {
+        if (txn.isActive()) {
+          txn.rollback();
+          throw new Error("Datastore Update has failed.");
+        }
       }
-    } else if (userResult.size() > 1) {
-      throw new IllegalArgumentException("More than one user with the userId was found.");
     }
+  }
 
+  public static GivrUser getUserById(String userId) {
+    PreparedQuery preparedQuery = getPreparedQueryResultOfUserWithProperty("userId", userId);
+    Entity entity = preparedQuery.asSingleEntity();
 
-    GivrUser user = new GivrUser(userId, isMaintainer, isLoggedIn, "" /* URL is not needed when User is logged in. */);
+    boolean isMaintainer = false;
+    boolean isLoggedIn = true;
+    String userEmail = "";
+
+    if (entity != null) {
+      isMaintainer = (boolean) entity.getProperty("isMaintainer");
+      userEmail = (String) entity.getProperty("userEmail");
+    }
+
+    GivrUser user = new GivrUser(userId, isMaintainer, isLoggedIn, "" /* URL is not needed when User is logged in. */, userEmail);
     return user;
   }
 
   public static GivrUser getUserByEmail(String email) {
     // TODO: Support OAuth.
-    String authDomain = "gmail.com";
-    User user = new User(email, authDomain);
-    String userId = user.getUserId();
 
-    return getUserByIdFromDatastore(userId);
+    if (!checkIfUserWithPropertyExists("userEmail", email)) {
+      return new GivrUser("", false, false, "", email);
+    }
+    
+    PreparedQuery preparedQuery = getPreparedQueryResultOfUserWithProperty("userEmail", email);
+    Entity entity = preparedQuery.asSingleEntity();
+
+    String userId = "";
+    boolean isMaintainer = false;
+    if (entity != null) {
+      userId = (String) entity.getProperty("userId");
+      isMaintainer = (boolean) entity.getProperty("isMaintainer");
+    }
+    return new GivrUser(userId, isMaintainer, false /* isLoggedIn */, "" /* loginURL */, email);
   }
 
-  public static GivrUser getLoggedInUser() {
+  public static GivrUser getCurrentLoggedInUser() {
```

Comment:
note that the email returned in the givruser is what's in the database, NOT what's in the current user!